Eliminate `with mock.patch.dict(os.environ, ...)` using GPT-4

The following code detects code blocks that contain `with mock.patch.dict(os.environ, ...)` and fix them:

<details><summary>Open</summary>
<p>

````python
"""
python a.py $(git ls-files tests/**/*.py)
"""
from __future__ import annotations

import random
import ast
import openai
import asyncio
import re
import textwrap
import click
import logging
from typing import Any, Iterator, TypeVar, Dict
from functools import total_ordering

T = TypeVar("T")

logger = logging.getLogger(__name__)


def get_indent(s: str) -> int:
    return len(s) - len(s.lstrip())


class Finder(ast.NodeVisitor):
    """
    Finds function nodes that contain a mock.patch.dict call
    """

    def __init__(self, filename):
        self.filename = filename
        self.funcs = []
        self.bad_funcs = []

    def node_to_location(self, node) -> str:
        return f"{self.filename}:{node.lineno}:{node.col_offset}"

    def visit_FunctionDef(self, node) -> None:
        self.funcs.append(node)
        super().generic_visit(node)
        self.funcs.pop()

    @property
    def current_func(self) -> ast.FunctionDef | None:
        if self.funcs:
            return self.funcs[-1]
        return None

    def visit_With(self, node) -> None:
        for item in node.items:
            call = item.context_expr
            if (
                isinstance(call, ast.Call)
                and isinstance(call.func, ast.Attribute)
                and call.func.attr == "dict"
                and isinstance(call.func.value, ast.Attribute)
                and call.func.value.attr == "patch"
                and isinstance(call.func.value.value, ast.Name)
                and call.func.value.value.id == "mock"
            ):
                for arg in call.args:
                    # Find os.environ
                    if (
                        isinstance(arg, ast.Attribute)
                        and arg.attr == "environ"
                        and isinstance(arg.value, ast.Name)
                        and arg.value.id == "os"
                    ) or (isinstance(arg, ast.Constant) and arg.value == "os.environ"):
                        print(self.node_to_location(self.funcs[-1]))
                        if self.current_func and self.current_func not in self.bad_funcs:
                            self.bad_funcs.append(self.funcs[-1])
                        break
        self.generic_visit(node)


def join_lines(lines: list[str]) -> str:
    return "\n".join(lines)


@total_ordering
class Range:
    def __init__(self, start: int, end: int) -> None:
        self.start = start
        self.end = end

    def offset(self, offset: int) -> Range:
        return Range(self.start + offset, self.end + offset)

    def __lt__(self, other: Range) -> bool:
        return self.start < other.start

    @classmethod
    def from_node(cls, node: ast.AST) -> Range:
        return cls(node.lineno - 1, node.end_lineno - 1)

    def slice(self, lines: list[str]) -> list[str]:
        return lines[self.start : self.end + 1]


class Patch:
    def __init__(self, rng: Range, lines: list[str]) -> None:
        self.range = rng
        self.lines = lines

    def diff(self) -> int:
        """
        How many lines does this patch add or remove if applied to a file?
        """
        return len(self.lines) - (self.range.end - self.range.start + 1)

    def offset(self, offset: int) -> Patch:
        return Patch(self.range.offset(offset), self.lines)

    def __repr__(self) -> str:
        c = join_lines(self.lines)
        return f"""\
lineno: {self.range.start}
end_lineno: {self.range.end}
-----
{c}
-----
"""

    def apply(self, lines: list[str]) -> list[str]:
        return [
            *lines[: self.range.start],
            *self.lines,
            *lines[(self.range.end + 1) :],
        ]

    @staticmethod
    def batch_apply(lines: list[str], patches: list[Patch]) -> list[str]:
        offset = 0
        for patch in patches:
            p = patch.offset(offset)
            lines = p.apply(lines)
            offset += patch.diff()
        return lines


def is_valid_code(code: str) -> bool:
    """
    Is `code` valid Python code?
    """
    try:
        ast.parse(textwrap.dedent(code))
        return True
    except SyntaxError:
        return False


class TopLevelImportFinder(ast.NodeVisitor):
    """
    Finds top-level import statements
    ```
    import a  # top-level
    import b  # top-level

    def foo():
        import c  # not top-level
    ```
    """

    def __init__(self):
        self.imports = []

    def visit_Import(self, node: ast.Import) -> None:
        self.imports.append(node)

    def visit_ImportFrom(self, node: ast.ImportFrom) -> None:
        self.imports.append(node)

    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
        pass


def find_imports(code: str) -> list[int]:
    finder = TopLevelImportFinder()
    finder.visit(ast.parse(textwrap.dedent(code)))
    line_nums = set()
    for imp in finder.imports:
        line_nums.update(range(imp.lineno, imp.end_lineno + 1))
    return sorted(line_nums)


def remove_imports(code: str) -> str:
    import_lines = find_imports(code)
    return join_lines(l for i, l in enumerate(code.split("\n")) if i + 1 not in import_lines)


def extract_code_from_block(s: str) -> str | None:
    if m := re.search(r"```(?:python)?\n(.+)\n```", s, re.DOTALL):
        return m.group(1)
    return None


async def chat_complete(func_lines: list[str]) -> str:
    prompt = """
The following pytest function contains `with mock.patch.dict(os.environ, dictionary, ...):`. We can use
`monkeypatch.setenv` or `monkepatch.delenv` instead. Please rewrite the function using `monkeypatch.setenv` or
`monkeypatch.delenv`. If the function arguments doesn't contain `monkeypatch`, please add it.
You might be tempted to add missing import statements. Please don't do that. You can just fix
 `with mock.patch.dict(os.environ, dictionary, ...):` and leave the rest of the code as is.

```python
{code}
```

The output should look like this:

```python
<new_code>
```
""".format(
        code=join_lines(func_lines)
    )
    resp = await openai.ChatCompletion.acreate(
        model="gpt-4",
        messages=[
            {"role": "user", "content": prompt},
        ],
        temperature=0.0,
    )
    return resp["choices"][0]["message"]["content"]


def extract_code(chat_resp: str) -> list[str] | None:
    logger.debug("Response: \n---\n%s\n---", chat_resp)
    new_code = extract_code_from_block(chat_resp)
    if new_code is None:
        return None
    if not is_valid_code(new_code):
        return None
    return new_code


def clean_up_code(code: str, indent: int) -> str:
    # When a function contains undefined modules, ChatGPT sometimes imports them even when
    # we don't ask it to. Those imports need to be removed.
    return remove_imports(textwrap.indent(code, " " * indent))


def batched(iterable: Iterator[T], n: int) -> Iterator[list[T]]:
    batch: list[T] = []
    for item in iterable:
        batch.append(item)
        if len(batch) == n:
            yield batch
            batch = []
    if batch:
        yield batch


def collect_funcs(files: list[str]) -> Dict[str, list[Range]]:
    funcs_by_files = {}
    for file in files:
        with open(file) as f:
            code = f.read()

        tree = ast.parse(code)
        visitor = Finder(file)
        visitor.visit(tree)

        for func in visitor.bad_funcs:
            funcs_by_files.setdefault(file, []).append(Range.from_node(func))
    return funcs_by_files


async def fix(files: list[str]) -> None:
    print("Collecting functions to fix")
    tasks_by_file = collect_funcs(files)
    print("=" * 30)
    print("Generating patches")
    for file, ranges in tasks_by_file.items():
        with open(file) as f:
            lines = f.read().split("\n")

        patches: list[Patch] = []
        for batch in batched(ranges, 10):
            for _ in range(3):  # Retry 3 times
                coros = (chat_complete(rng.slice(lines)) for rng in batch)
                results = await asyncio.gather(*coros, return_exceptions=True)
                failed: list[Range] = []
                for rng, resp in zip(batch, results):
                    if isinstance(resp, Exception):
                        failed.append(rng)
                        logger.error("Failed to generate patch for %s", rng)
                        continue

                    # print(resp)
                    # code = resp
                    if code := extract_code(resp):
                        indent = get_indent(rng.slice(lines)[0])
                        code = textwrap.indent(code, " " * indent)
                        code = remove_imports(code)
                        patches.append(Patch(rng, code.split("\n")))

                if not failed:
                    break
                batch = failed

        patches.sort(key=lambda p: p.range)
        print(patches)
        new_lines = Patch.batch_apply(lines, patches)
        with open(file, "w") as f:
            f.write(join_lines(new_lines))

        break  # comment out this line to fix all files


@click.command()
@click.option("--debug", is_flag=True, default=False, help="Turn on debug logging")
@click.argument("files", nargs=-1)
def main(debug: bool, files: tuple[str]) -> None:
    if debug:
        logging.basicConfig(level=logging.DEBUG)
    files = list(files)
    random.shuffle(files)
    asyncio.run(fix(files))


if __name__ == "__main__":
    main()

````

</p>
</details> 


## Example

```python
def test_foo():
    with mock.patch.dict(os.environ, ...):
        ...
```

and rewrites it using pytest's `monkeypatch` fixture using GPT-4:

```python
def test_foo(monkeypatch):
    monkeypatch.setenv(...)
```


## Task

1. Run `export OPENAI_API_KEY=...` to set the API key.
1. Save the code as `a.py` in the repository root.
1. Run `python a.py $(git ls-files tests/**/*.py)`. **To make the PR reviewable, the script randomly chooses a file and only fixes that file**.
1. CAREFULLY verify the new code looks correct. AI is not always accurate.
1. Make manual fixes if necessary.
1. File a PR. Make sure the PR only fixes a single file.

Hi @harupy @gabrielfu! 

Following [PR1](https://github.com/mlflow/mlflow/pull/9086#issuecomment-1648185698) I have  changed [tests/projects/test_docker_projects.py](https://github.com/mlflow/mlflow/blob/13f355577cf925afefae1cd246d47a8a3517ffe1/tests/projects/test_docker_projects.py#L262)
and faced pylint restriction:
`pylint tests/projects/test_docker_projects.py` gives me:

`
tests/projects/test_docker_projects.py:261:8: W0004: with pytest.raises should not contain multiple statements. It should only contain a single statement that throws an exception. (pytest-raises-multiple-statements)
` 
This is because this [custom plugin](https://github.com/mlflow/mlflow/blob/13f355577cf925afefae1cd246d47a8a3517ffe1/pylint_plugins/pytest_raises_checker/__init__.py#L40) doesn't allow to make commit.
 
However with `with mock.patch.dict("os.environ", os_environ):` pylint works fine.

Waiting for further instructions how to solve this part)

Thank you

--- a/tests/store/artifact/test_http_artifact_repo.py
+++ b/tests/store/artifact/test_http_artifact_repo.py
@@ -261,7 +261,7 @@ def test_download_artifacts(http_artifact_repo, tmp_path):
         assert read_file(paths[1]) == "data_b"
 
 
-def test_default_host_creds():
+def test_default_host_creds(monkeypatch):
     artifact_uri = "https://test.com"
     username = "user"
     password = "pass"
@@ -282,8 +282,7 @@ def test_default_host_creds():
 
     repo = HttpArtifactRepository(artifact_uri)
 
-    with mock.patch.dict(
-        "mlflow.tracking._tracking_service.utils.os.environ",
+    monkeypatch.setenvs(
         {
             MLFLOW_TRACKING_USERNAME.name: username,
             MLFLOW_TRACKING_PASSWORD.name: password,
@@ -291,9 +290,9 @@ def test_default_host_creds():
             MLFLOW_TRACKING_INSECURE_TLS.name: str(ignore_tls_verification),
             MLFLOW_TRACKING_CLIENT_CERT_PATH.name: client_cert_path,
             MLFLOW_TRACKING_SERVER_CERT_PATH.name: server_cert_path,
-        },
-    ):
-        assert repo._host_creds == expected_host_creds
+        }
+    )
+    assert repo._host_creds == expected_host_creds
 
 
 @pytest.mark.parametrize("remote_file_path", ["a.txt", "dir/b.txt", None])

--- a/tests/store/artifact/test_s3_artifact_repo.py
+++ b/tests/store/artifact/test_s3_artifact_repo.py
@@ -66,7 +66,7 @@ def test_file_artifact_is_logged_with_content_metadata(s3_artifact_root, tmp_pat
     assert response.get("ContentEncoding") is None
 
 
-def test_get_s3_client_hits_cache(s3_artifact_root):
+def test_get_s3_client_hits_cache(s3_artifact_root, monkeypatch):
     # pylint: disable=no-value-for-parameter
     repo = get_artifact_repository(posixpath.join(s3_artifact_root, "some/path"))
     repo._get_s3_client()
@@ -81,12 +81,8 @@ def test_get_s3_client_hits_cache(s3_artifact_root):
     assert cache_info.misses == 1
     assert cache_info.currsize == 1
 
-    with mock.patch.dict(
-        "os.environ",
-        {"MLFLOW_EXPERIMENTAL_S3_SIGNATURE_VERSION": "s3v2"},
-        clear=True,
-    ):
-        repo._get_s3_client()
+    monkeypatch.setenv("MLFLOW_EXPERIMENTAL_S3_SIGNATURE_VERSION", "s3v2")
+    repo._get_s3_client()
     cache_info = _cached_get_s3_client.cache_info()
     assert cache_info.hits == 1
     assert cache_info.misses == 2

--- a/tests/utils/test_rest_utils.py
+++ b/tests/utils/test_rest_utils.py
@@ -1,4 +1,3 @@
-import os
 from unittest import mock
 import re
 import numpy

--- a/tests/utils/test_rest_utils.py
+++ b/tests/utils/test_rest_utils.py
@@ -157,19 +157,18 @@ def test_http_request_with_basic_auth(request):
 
 
 @mock.patch("requests.Session.request")
-@mock.patch.dict(
-    os.environ,
-    {
-        "AWS_ACCESS_KEY_ID": "access-key",
-        "AWS_SECRET_ACCESS_KEY": "secret-key",
-        "AWS_DEFAULT_REGION": "eu-west-1",
-    },
-)
-def test_http_request_with_aws_sigv4(request):
+def test_http_request_with_aws_sigv4(request, monkeypatch):
     """This test requires the "requests_auth_aws_sigv4" package to be installed"""
 
     from requests_auth_aws_sigv4 import AWSSigV4
 
+    monkeypatch.setenvs(
+        {
+            "AWS_ACCESS_KEY_ID": "access-key",
+            "AWS_SECRET_ACCESS_KEY": "secret-key",
+            "AWS_DEFAULT_REGION": "eu-west-1",
+        }
+    )
     aws_sigv4 = MlflowHostCreds("http://my-host", aws_sigv4=True)
     response = mock.MagicMock()
     response.status_code = 200

